%*****************************************
\chapter{Introduction to RNA folding}\label{ch:folding}
The function of non-coding RNAs is largely determined by their three-dimensional structure \cite{cech2014noncoding}. For instance, we can analyze the catalytic function of ribozymes in terms of basic structural motifs, e.g. hammerhead or hairpin structures \cite{doherty2001ribozyme}. Other RNAs, like riboswitches, involve changes between alternative structures  \cite{vitreschak04_ribos}. Therefore, understanding the relation sequence and structure is a central challenge in molecular biology. In the last $20$ years, many different methods for determining the RNA structures of molecules have emerged: from experimental lab methods to computational approaches. For experimental lab methods, X-ray crystallography and the nuclear magnetic resonance (NMR)  are the most accurate approaches to offer structural information at a single base-pair resolution. Both experimental methods are often characterized by high experimental cost and low throughput. In addition to those limitations, RNA molecules are volatile and difficult to crystallize. Despite the development of more sophisticated techniques to infer the state of nucleotides in RNA molecules using enzymatic \cite{kertesz2010genome, underwood2010fragseq} or chemical probes \cite{tijerina2007dms, wilkinson2006selective} coupled with next-generation sequencing \cite{bevilacqua2016genome, tian2016rna}, most of them can only capture RNA structures \textit{in vitro} which mostly differ from the \textit{in vivo} structure conformations. Experimentally, only a tiny fraction of known ncRNAs has been determined \cite{rnacentral2017rnacentral}. Because measuring the structure of RNAs experimentally is very difficult and expensive, computational approaches play a central role in the analysis of natural RNAs \cite{seetin2012rna, fallmann2017recent}, and are an important alternative to experimental approaches. In nature, RNAs fold into secondary structures before folding into higher-level (tertiary and quaternary) structures \cite{brion1997hierarchy,tinoco1999rna}. This separation of time scales justifies focusing on the prediction of secondary structure; evidence suggests that the resulting high-level structures are indeed largely determined by the RNA's secondary structures. This chapter provides an overview of computational folding methods of RNA molecules into a secondary structure. Two techniques will be reviewed: statistical approaches such as machine learning and score-based methods.

\section{Stability and prediction of RNA secondary structures}
%\graffito{Note: The content of this chapter is just some dummy text.
%It is not a real language.}

The mapping from RNA sequences to their corresponding secondary structure defines the folding of RNA molecules. RNA folding is, therefore, a process by which a linear RNA sequence acquires a secondary structure through intra-molecular interactions. The nature of those interactions defines the thermodynamic stability of the secondary structure. The thermodynamic stability $\Delta G_{\sigma}$ of a structure $\sigma$ is the free energy difference with respect to the completely unfolded state. Therefore, the free energy function defines the mapping of an RNA sequence to its corresponding free energy.
Most computational methods search for structures that minimize this free energy function to predict biologically relevant structures. Structures are decomposed into components called loops to compute the free energy (See Definition 4). The loop decomposition allows building the basis of the standard energy model for RNA secondary structures called the \textit{nearest neighbour} model \cite{turner09_nndb}. This model associates tabulated free energy values to loop types and nucleotide compositions; the \texttt{Turner2004} \cite{mathews2004incorporating}  and the \texttt{Turner1999} \cite{mathews1999expanded} are the most widely used parameter sets.  The total free energy of a secondary structure is assumed to be a sum over its constituent loops according to the additivity principle \cite{dill97_addit_princ_bioch}. This structure decomposition allows an efficient dynamic programming (DP) algorithm to determine the minimum free energy (MFE) pseudoknot-free structure of a sequence $\phi$ in the structure space $\Sigma_{\phi}$. 

The DP technique is one of the most widely used score-based methods. The first DP algorithm that finds the structure with the maximum base pairs was proposed by Nussinov, and Jacobson \cite{nussinov1980fast}. A few years after, Zucker and Stieger \cite{zuker1981optimal} extended Nussinov's algorithm to a more realistic scoring model based on free energy, the NN model. Almost all score-based methods rely on the same DP algorithm, but the decomposition scheme and the scoring model could defer from one to another. When predicting structures with non-canonical base pairs, some other scoring schemes are used, such as nucleotide cyclic motifs score system \cite{zu2011folding,parisien2008mc,dallaire2016exploring} or equilibrium partition function \cite{sloma2017base}. Besides the score-based methods, we have the comparative sequence analysis methods, the most computationally accurate for determining the RNA secondary structure \cite{gutell2002accuracy, madison1966nucleotide}. Using the set of homologous structures, This method allows finding base pairs that covary to maintain WC and wobble bases of a given sequence $\phi$ \cite{gutell1985comparative}. The first comparative method predicting a common secondary structure conserved in the given homologous sequence set was developed by Han and Kim in the early nineteenth, and it was based on the comparative phylogenetic analysis. When neglecting the special base pairs and the weak interactions, the running time of both approaches (sore-based and comparative analysis) is usually $O(L^3)$ (Where $L$ is the RNA sequence length). Many other comparative analysis methods and variations of score-based methods were also proposed to improve the computational time. More recently, a heuristic method such as \texttt{LinearFold} allows achieving good RNA folding performance in a linear time ($O(L)$). 

When pseudoknots are considered, the loop decomposition of a secondary structure and the energy rules break down. Although we can assign reasonable free energies to the helices in a pseudoknot and even to possible coaxial stacking between them, it is impossible to estimate the effects of the new kinds of loops created. Base triples pose an even greater challenge because the exact nature of the triple cannot be predicted in advance, and even if it could, we have no data for assigning free energies.
Nevertheless, there are existing techniques that approximate the energies of pseudoknot loops and allow the dynamic programming technique to tackle the RNA folding with pseudoknots. However, the time complexity still reminds the main problem. Using the DP technique for the pseudoknot structure prediction, the time complexity goes up to $O(L^6)$ for the exact prediction. But for heuristic methods such as \texttt{IPKnot} \cite{sato2011ipknot} and \texttt{Hotknots} \cite{ren2005hotknots}, the running time can be reduced down to $O(L^4)$. 

Despite the advanced development of computational tools for RNA folding, it's challenging to understand the folding mechanism fully. In contrast to score-based and comparative analysis methods, machine learning methods are data-driven methods that require no knowledge of the folding mechanism. Nevertheless, the requirement of ML-based methods is a large amount of training data on which they can learn. In the last few decades, ML methods have been used for many aspects of RNA secondary structure prediction methods to improve the prediction performance and overcome the limitations of existing methods. However, they did not replace the mainstream score-based methods with respect to accuracy and generalization. In addition to some overfitting concerns, ML-based methods cannot give dynamic information on the RNA folding process since little data are available on structural dynamics. In addition, the training data used in ML-based methods are mostly obtained through phylogenetic analyses. Consequently, their prediction may be biased due to the \textit{in vivo} third elements. The following subsections provide a detailed description of some of the recent ML-based and score-based tools for secondary structure prediction.
\subsection{ MFE prediction tools for pseudoknot-free RNA sequences using a score-base method}
The score-based methods often assume that the native or biological RNA structure is the one that minimizes/maximizes the overall total score, depending on the hypotheses made on the RNA folding mechanism. In the pseudoknot-free MFE prediction, where the special and weak interactions are neglected, the folding problem is less complex, and the scoring model is the free energy. Hence, the issue of RNA secondary structure prediction becomes an optimization problem that aims at finding the best-scoring structure $\mathcal{S}^{MFE}$ by minimizing a scoring function $\Delta G$.

\begin{equation}
	\mathcal{S}^{MFE} =  \textbf{argmin}_{\mathcal{S} \in \Sigma_{phi}} \Delta G(\mathcal{S}, \phi) 
\end{equation}

Where $\Sigma_{\phi}$ is the set of all possible pseudo-knot free secondary structure for the sequence $\phi$ of lenght $L$ and, $\Delta G(\mathcal{S}, \phi)$ the free energy of the structure $\mathcal{S}$ evaluated for the sequence $\phi$.

Since each possible structure can be uniquely and recursively decomposed into smaller components (or loops) with independent free energy contributions, the DP is best suited for most of the following tools presented here.
%Equation \textcolor{red}{4(to fill up)}  shows the scoring function used in the DP tools presented in this work.

\begin{itemize}
	\item \texttt{Unfold} \cite{zuker1981optimal, zuker1984rna}: It is the successor of the original  \texttt{mfold} program which was the first realistic implementation of the DP for secondary structure predictions with a score based on the loop energy parameters and a worse case time complexity of $O(L^3)$. The initial version was an improvement of the simplest DP for secondary structure prediction known as the \textit{maximum circular matching problem}. The authors demonstrated that the loop-based energy model is also amenable to the same algorithmic ideas. With McCaskill's algorithm \cite{mccaskill1990equilibrium}, for computing the partition function of the equilibrium ensemble of RNA molecules, more efficient implementations of the initial program with accurate thermodynamic modelling have been provided. The latest implementation is known as \texttt{Unfold}.
	
	\item \texttt{RNAStructure} \cite{matthews1998updated,reuter2010rnastructure}: The software first appeared in 1998 as a reimplementation of the program \texttt{mfold} with improved thermodynamic parameters. In its initial version, four major changes were made in \texttt{mfold}: (1) an improvement on the methods for forcing base pairs; (2) a filter that removed isolated WC or wobble base pairs has been added; (3) the energy parameter for interior, internal and hairpin loops were incorporated; (4) a new model for coaxial stacking of helices. It predicts the lowest free energy structure and a set of low energy structures. The new implementation also provided a user-friendly graphical interface for Windows operating system. Subsequently, the first implementation was extended to include biomolecular folding; an algorithm that finds low free energy structures common to two sequences; the partition function algorithm and all free energy structures, and the constraints with enzymatic data and chemical mapping data. The recent version includes the partition function computation for secondary structures common to two sequences and can perform stochastic sampling of common structures \cite{harmanci2009stochastic}. Additionally, it contains \texttt{MaxExpect}, which finds maximum expected accuracy structures \cite{lu2009improved}, and a method for removal of pseudoknots, leaving behind the lowest free energy pseudoknot-free structure. 
	
	\item \texttt{RNAfold} \cite{hofacker1994fast,lorenz11_vienn_packag}: It is one of the most used and efficient folding tools. It computes the MFE secondary structure using an efficient DP scheme and backtraces an optimum structure. It also allows computing the partition function using McCaskill's algorithm, the matrix of base pairing probabilities, and the centroid structure. It is part of the \texttt{ViennaRNA Package}. Since its first version, it aims at suggesting an efficient implementation of Zucker's algorithm with more flexibility on the folding constraints. Many other versions have been released, including a GPU implementation. The latest stable release of the \texttt{ViennaRNA Package} is Version 2.5.0.
	
	%\item \texttt{RNAshape}\cite{steffen2006rnashapes}:
	\item \texttt{LinearFold} \cite{huang2019linearfold}: For many decades, the DP techniques have been the most accurate and fast at predicting pseudoknot-free structure for short input RNA sequences. But for long sequences, the prediction remains challenging because of the computational time and the lack of accurate thermodynamic energy parameters. In contrast to traditional DP methods which are often bottom-up, \texttt{LinearFold} is a left-to-right DP. The left-to-right DP consists of scanning the input RNA sequence $\phi$ from left to right,  maintaining a \textit{stack} along the way and performing one of the three actions (\textit{push, skip} or \textit{pop}). The \textit{stack} consists of a list of unpaired opening bracket positions and at each position $j = 1\dots L$, the three actions consist respectively of 1)push: opening a bracket at position $j$, 2) skip: unpaired nucleotide at position $j$ and 3) pop: closing the bracket at position $j$. Initially, \texttt{LinearFold}'s computational time was similar to the classical DP ($O(L^3)$) because of the \textit{pop} action that involves three free indices (i.e. unpaired positions). But using a beam search heuristic, the time complexity was then reduced to $O(Lb\log  b)$, where $b$ is the beam size. The beam search is a popular heuristic technique used in computational linguistics. This technique allows keeping only the top $b$ highest-scoring (or low energy) states for each prefix of the input sequences.
\end{itemize}

%Beside the above-mentioned tools there are others score-based methods that do not implement a DP algorithm: 
Although the sore-based approaches for RNA structure prediction often offer good accuracy and generalization,  the non-availability of the thermodynamic energy parameters for specific loops of extended sizes presents the main challenge for predicting long sequences (i.e. $L \geq 1,000$ nucleotides).  Early ML-based methods aim to improve the energy parameters by learning the underlying folding patterns from a more considerable amount of training data.  In the next section of this chapter, we will present some of the recent improvements in structure prediction using ML-based methods.
\subsection{ML-based methods}
The ML-based methods for RNA secondary structure prediction can generally be classified into three categories according to ML's subprocess, i.e., score scheme based on ML, preprocessing and postprocessing based on ML, and prediction process on ML. All the ML-based methods in these three categories trained their models in a supervised way \cite{zhao2021review}. 

When using a scoring scheme based on ML,  the parameter estimation in the scoring scheme is first optimized using an ML model. The estimated parameters are then used to evaluate the scores of possible conformations. Difference scoring schemes can be refined by using that approach: the free energy parameters, weights, and probabilities. The free energy parameter-refining is the most popular because several thermodynamic parameters of the NN model have to be based on a large number of optimal melting experiments and the experiments are time and labour-consuming. In fact, not all free energy changes in structural elements can be experimentally measured because of technical difficulties. Instead of refining the free energy parameters, some ML-based approaches scream through existing data of RNA structures to extract weights that consist of different features of RNA structure elements. These weights can be used as a scoring function for DP techniques. The advantage of such a scoring function is that it decouples structure prediction and energy estimation. However, learned weights have no explanations because of the ML black box. 

Another alternative for predicting RNA structures is the stochastic context-free grammars (SCFG) \cite{sakakibara1994stochastic, rivas2012range, dowell2004evaluation, knudsen1999rna, knudsen2003pfold, woodson2000recent}. SCFGs allow building grammar rules and induce a join probability distribution over possible RNA structure for a given sequence $\phi$. In addition, the SCFG models specify probability parameters for each production rule in the grammar, which allow assigning a probability to each sequence generated by the grammar. These probability parameters are learned from datasets of RNA sequences associated with known secondary structures without carrying any external laboratory experiments \cite{dowell2004evaluation}. 

Besides the ML-based methods that focus on refining the folding parameters, there are preprocessing and post-processing based on ML \cite{hor2013tool, zhu2018research,haynes2008using} and direct predicting process based on ML \cite{takefuji1990parallel,liu2006hopfield,steeg1993neural}. Preprocessing and postprocessing models allow for choosing the appropriate prediction method or set of prediction parameter sets and provide a means of determining the most likely structures among the possible outcomes that are useful for decision. The preprocessing and postprocessing ML tools are often based on a support vector machine (SVM). 

Finally, it is possible to use ML techniques to predict RNA secondary structure directly or combine it with other algorithms in an end-to-end fashion. Below are some of the most used and recent ML-based tools for RNA secondary structure prediction.
\begin{itemize}
	\item \texttt{ContraFold}\cite{do2006contrafold}: Using the so-called probabilistic model, the conditional log-linear model (CLLM), \texttt{ContraFold} appeared early in 2006 as the first  probabilistic prediction tool outperforming all the existing tools including the thermodynamic tools such as \texttt{RNAfold} and \texttt{mfold}. The CLLMs is a flexible class of probabilistic models which generalize upon SCFGs by using discriminative training and feature-rich scoring. \texttt{ContraFold} implements a CLLM incorporating most of the features found in typical thermodynamic models allowing the tool to achieve the highest single sequence prediction accuracy to date, when compared with the currently available probabilistic models.
	\item \texttt{ContextFold} \cite{zakov2011rich}: In contrast to \texttt{ContraFold}, \texttt{ContextFold} utilizes a weighted appraoch based on ML. In particular, it uses a discriminative structured-prediction learning framework combined to online learning algorithm. \texttt{ContextFold} uses a large training dataset of RNA sequences annotated with theirs corresponding structures to obtain a ML model made of $70,000$ free parameters which has several orders of magnitudes compared to traditional models (i.e. thermodynamic free energy parameters). At it first apparition, \texttt{ContextFold}'s model succeeded at the error reduction of about $50\%$ but to date, some over fitting concerns have been reported when using the tool, especially for the prediction of structures with large unpaired regions.  
	\item \texttt{Mxfold2}  \cite{sato2021rna}: It is one of the most recent ML-based tools for predicting the secondary structure of RNA molecule. Its particularity is the ML technique used, a Deep Neural Network (DNN). it also belongs to the weighted approach based on ML since, the resulting model of a DNN is a set of weight parameters. \texttt{MxFold2}'s DNN uses the max-margin framework with thermodynamic regularization, which made the folding scores predicted by \texttt{MXfold2} and the free energy calculated by the thermodynamic parameters closer as possible. This method has shown robust prediction on both sequences and families of natural RNAs, which suggests that the weighted ML approaches can compensate the gaps in the thermodynamic parameter approaches. 
\end{itemize}

\subsection{Prediction tools for pseudoknotted RNA sequences}
Folding RNA sequences with pseudoknotted interactions is  computationally more expensive than a pseudoknot-free target because of the time complexity of the folding algorithms. Specifically, the time complexity of the pseudoknot-free secondary structure prediction is $O(L^3)$ when using dynamic programming approaches such as \texttt{RNAfold},  or less with heuristic folding methods (e.g. $O(L)$ for \texttt{LinearFold} and $O(L^2\log L)$). By contrast, when considering a special class of pseudoknots, the time complexity of folding goes up to $O(L^6)$ for an exact thermodynamic prediction using a dynamic programming approach such as \cite{pseudoknotDP}. When Using heuristic methods, the time complexity slows down to $O(L^4)$ (e.g. tools such as \texttt{IPknot} and \texttt{HotKnots}) or $O(L^3)$ for tool such as \texttt{HFold}.  In this section, we introduction couple of tools for predicting RNA pseudoknotted structures.
\begin{itemize}
	\item \texttt{pKiss} \cite{jangie2015}: The program \texttt{pKiss} appears the first time in 2014 as an updated version of the program \texttt{pknotsRG}\cite{reeder2007pknotsrg} which is a module of the RNA abstract shapes analysis  \texttt{RNAshapes} \cite{jangie2015}. Initially, the program \texttt{pknotsRG} was build for the prediction of some special class of pseudoknots (unknotted structures and H-type pseudoknots) and then, it was extended to predict RNA structures that exhibit kissing hairpin motifs in an arbitrarily nested fashion, requiring $O(L^4)$ time.  In addition to predicting the kissing hairpin motifs, \texttt{pKiss} also provides new features such as: shape analysis, computation of probabilities, different folding strategies
	and different dangling base models. 
	\item \texttt{IPknot \cite{sato2011ipknot}}: it was first introduced in a paper by Kengo and his collaborators in 2011 as a novel computational tool for predicting RNA secondary structure with pseudoknots using integer programming technique. \texttt{IPknot} uses the maximum expected accuracy (MEA) as scoring function and the maximizing expected accuracy problem is solved using an integer programming with threshold cut. In fact, \texttt{IPknot} decomposes a pseudoknotted  structure into a set of pseudoknot-free substructures and approximates a base-pairing probability distribution that considers pseudoknots, leading to the capability of modeling a wide class of pseudoknots and running quite fast. In addition to single sequence analysis, \texttt{IPknot} can also predict the consensus secondary structure with pseudoknots when a multiple sequence alignment is given.
	\item \texttt{HotKnots} \cite{ren2005hotknots}. In contrast to the previously mentioned tools, \texttt{HotKnots} implements an heuristic algorithm based on the simple idea of iteratively forming stable stems. By iteratively forming stems, the algorithm explore many alternative secondary structures using a free energy minimization for pseudoknot-free secondary structures. For each structure formed at each step, several different additions of a single substructure are considered, resulting in a tree of candidate structures. the criterion for determining which substructures to add to partially formed structures at successive levels of the tree was also new, relative to previous algorithms: energetically favorable substructures called ‘‘hotspots’’ are found by a call to Zuker’s algorithm, with the constraint that no base already paired may be in the structure.
\end{itemize}

\section{RNA kinetics} % \ensuremath{\NoCaseChange{\mathbb{ZNR}}}
The previous section introduced how secondary structures with theirs thermodynamic properties can be predicted. However, the methods used for predictions do not tell us nothing about how the structures change overtime and how they are related to each other. In the folling section, we will discuss the folding dynamics of RNA molecule. 

The folding of RNA molecules is a remarkably more complex. It is a result of the delicate balance between multiple factors: the chain entropy, ion-mediated electrostatic interactions and solvation effect, base pairing and stacking, and other non-canonical interactions \cite{chen2008rna}. It is a dynamic process which is governed by is a constant formation and/or dissolving of base pairs. In other terms, the RNA molecule navigates its structure space by following a free energy landscape. The free energy landscape here is a high-dimensional space of all possible secondary structures ($\Gamma$) weighted by their free energy $E(\Gamma)$.  


As usually done, the kinetics is modelled as a continuous-time Markov chain \cite{lorenz20_effic_comput_base_probab_multi_rna_foldin}, where populations of structure evolve according to transition rates. In this context, an Arrhenius formulation 
\begin{equation}
	\label{Eq:arrhenius}
	k_{i \rightarrow j} = k_0 \text{exp}(-\beta  \Delta E^{\ddagger}_{i\rightarrow j})
\end{equation}
is commonly used to derive elementary transition from state $i$ to state $j$; where \(\Delta E^{\ddagger}_{i \rightarrow j}\) is the activation barrier separating \(i\) from \(j\), and \(\beta=1/k_BT\) is the inverse thermal energy (mol/kcal). 
%In contrast, our kinetic ansatz uses transition rates \(r(x\rightarrow y)\) based on the Metropolis scheme already used in \cite{klemm2008funnels}, and defined as
%\begin{equation}
%\label{Eq:metropolis}
%r(x\rightarrow y) = k_0 \times \text{min}(1, \text{exp}(-\beta \Delta \Delta G(x\rightarrow y))),
%\end{equation}
%where \(\Delta \Delta G(x\rightarrow y)\) is the stability change between structure \(x\) and \(y\). 
Here \(k_0\) is the fundamental rate constant, which is solvent dependent.
Three rate models describing elementary steps in the structure space are often used to study RNAs folding dynamics: 
\begin{enumerate}
	\item The base stack model \cite{zhang02_rna_hairp_foldin_kinet,zhang2003analyzing,zhang2006exploring}: it uses base stacks as elementary kinetic move. A move consists of an addition or a breaking of a base stack with \(\Delta E^{\ddagger}_{i \rightarrow j}\) equal to the change in the entropic free energy $T\Delta S$ and the enthalpy $\Delta H$, respectively.  
	\item The base pair model \cite{flamm2000rna,cocco2003slow}: It uses base pair as elementary kinetic steps which gives the finest resolution, but at the cost of computation time. Here \(\Delta E^{\ddagger}_{i \rightarrow j}= \Delta E/2\) where $\Delta E$ is the energy change from state $i$ to state $j$ or $ \Delta E^{\ddagger}_{i \rightarrow j}= \Delta E$ for $\Delta E \geq 0$.
	\item  The helix stem model \cite{martinez84_rna_foldin_rule, isambert2000modeling}:  the elementary move is the creation or deletion of a helix stem. It provides a coarse-grained description of the dynamics where free energy changes (\(\Delta E^{\ddagger}_{i \rightarrow j}\)) due to stem formation guide the folding process. 
\end{enumerate}

The different rate models can lead to different folding pathways. The key factor that distinguishes the different rate models is whether the barrier is determined by ($\Delta 􏰦H, 􏰦\Delta S$) or by $\Delta E$. The  ($\Delta 􏰦H, 􏰦\Delta S$) values for different RNA base stacks show well-separated discrete hierarchies, whereas the $􏰦\Delta E$ values show no such large separation. For two typical base stacks, 5′AU-AU3′ and 5′UC-GA3', the difference $ \Delta 􏰦(􏰦\Delta H_{stact}, \Delta 􏰦S_{stack}) = (7.4 \text{ kcal/mol}$, $20 \text{ kcal/mol})$ is much larger than the difference  $\Delta (\Delta E_{stack})=1.4 \text{kcal/mol}$\cite{serra199511}. Because of this fact, different models can give different folding kinetics.

Depending on the rate model used, the following master-equation describe the population kinetics $p_i(t)$ for the $i^{th}$ state ($i=1\dots \Omega$, where $\Omega$ is the total number of chain conformations).

\begin{equation}
\label{Eq:kenetics}
\frac{\text{d}p_i(t)}{\text{d}t} = \sum\limits_{j \in \Omega}
k_{j \rightarrow i} p_{j}(t) - k_{i \rightarrow j} p_{i}(t),
\end{equation}

where $k_{j\rightarrow i}$ and $k_{i \rightarrow j}$ are the rate constants for the respective transitions. The equivalent matrix form of Equation \ref{Eq:kenetics} is given by: 
\begin{equation}
\label{Eq:matrixkinetic}
	\frac{\text{d}\textbf{p}(t)}{\text{d}t} = \textbf{M} . \textbf{p},
\end{equation}
where $\textbf{p} = (p_i, \dots p_{\Omega})$ is a column vector representing the frequency of structure at state $(i, \dots , \Omega)$ and, \textbf{M} is the rate matrix defined as: 

\begin{equation}
	\textbf{M}_{ij} =  \begin{cases}
	k_{i\rightarrow j},& \text{if } i\neq j\\
	- \sum_{j\neq i}{k_{ij}},              & \text{if } i=j
	\end{cases}
\end{equation}

For a given initial folding condition $p_i(0)$, the Equation \ref{Eq:matrixkinetic} is solvable by diagonalizing the rate matrix \textbf{M} and, the solution is the population kinetics $\textbf{p}(t)$ for $t>0$ is given by: 

\begin{equation}
	\label{Eq:solutionkinetics}
	\textbf{p}(t) = \sum_{m=1}^{\Omega} {C_m \textbf{n}_m \exp{-\lambda_m t}}
\end{equation}

where $-\lambda$ and $\textbf{n}_m$ are the $m^{th}$ eigenvalue and eigenvector of the rate matrix \textbf{M}, and $C_m$ is the coefficient that is dependent on the initial condition. The eigenvalue spectrum gives the rates of the kinetic modes of the system.

Simulating the RNA dynamics using the Equation \ref{Eq:kenetics} has some limitations. The solution to the master-equation given by Equation \ref{Eq:solutionkinetics} can only give ensemble-average macroscopic kinetics and cannot give detailed information about the microscopic pathways \cite{zhang06_explor_compl_foldin_kinet_rna_hairp}. Moreover the number of structures ($\Omega$) increases rapidly with the RNA sequence length $L$.  Therefore, the master equation is often limited to short RNA sequences. Because of these limitations a kinetics-cluster methods are alternatively used. The basic idea of the kinetic-cluster method is to classify the large structural ensemble into a much reduced systems of cluster (of macrostates) such that the overall kinetics can be represented by the inter-cluster transitions. Although both the master-equation and the kinetic-cluster methods can predict the macroscopic kinetics, and both are based on the complete conformational ensemble, the kinetic-cluster approach has the unique advantage of providing the direct information on the microscopic pathway statistics from the inter-cluster transitions \cite{zhang06_explor_compl_foldin_kinet_rna_hairp}. An alternative approach, implemented in \texttt{kinwalker} \cite{geis2008folding}, used the observation that folded intermediates are generally locally optimal conformations.  
\\
Although the above mentioned theoretical models allow to simulate the dynamics of RNA folding molecules they often missed an important component, namely, the sequence-dependent conformational statics of the single-stranded coil and loop states. This important component has been suggested by several experimental studies \cite{nagel2006structural, bonnet1998kinetics}. 
In folding experiments, Pan and coworkers observed two kinds of pathways in the free energy landscape of a natural ribozyme \cite{pan97_foldin_rna_invol_paral_pathw}. Firstly, the experiments revealed fast-folding pathways, in which a sub-population of RNAs folded rapidly into the native state. The second population, however, quickly reached metastable misfolded states, then slowly folded into the native structure. In some cases, these metastable states are functional. These phenomena are direct consequences of the rugged nature of the RNA folding landscape \cite{solomatin10_multip_nativ_states_reveal_persis}. The experiments performed by Russell and coworkers also revealed the presence of multiple deep channels separated by high energy barriers on the folding landscape, leading to fast and slow folding pathways \cite{russell2002exploring}. The formal description of the above mechanism, called kinetic partitioning mechanism, was first introduced by Guo and Thirumalai in the context of protein folding \cite{guo95_kinet_protein_foldin}. In the free energy landscape, these metastable conformations form competing attraction basins in which RNA molecules are temporarily trapped. However, \textit{in vivo}, folding into the native states can be promoted by molecular chaperones \cite{chakrabarti2017molecular}, which means that the active structure depends on factors other than the sequence. This may rise some discrepancy when comparing thermodynamic modelling to real data and the experimental verification of the rate model is also a challenge because the microscopic elementary processes are hidden in the ensemble averages of the measured kinetics.
Many researchers believe that single-molecule experiments, with careful extrapolation to the force-free case, may provide a discerning measure. All-atom simulations with reliable force field and sampling method are highly valuable for providing detailed atomistic configurations for the transition state \cite{chen2008rna}. Alternatively, systematic theory- experiment tests as done in \cite{zhang06_explor_compl_foldin_kinet_rna_hairp} for designed sequences can also provide critical assessment for the different rate models.


\section{Conclusion}
In this chapter, we have presented the RNA folding in two main steps: (1) the prediction of the secondary structure of RNA which represents the static part of the folding process; (2) the RNA kinetics which  aim at modelling the dynamics of the folding. The prediction of RNA secondary structure was introduced as an optimization problem and a review on existing methods and tools have been presented. In the next chapter, we will introduce the first result of our thesis that aims at predicting efficiently RNA folding pathways using the fast Fourier transform. The predicted pathways will then allow us to derive a set of energetically suboptimal structures from which we will model the slow folding process of RNA molecules.
%*****************************************
%*****************************************
%*****************************************
%*****************************************
%*****************************************
